pkg/basinprovider/provider.go

```diff
@@ -132,14 +135,79 @@ func (bp *BasinProvider) Push(ctx context.Context, ns string, rel string, tx bas
 	return err
 }
 
+// Upload uploads a file to th server.
+func (bp *BasinProvider) Upload(
+	ctx context.Context, ns string, rel string, r io.Reader, signer *app.Signer, progress io.Writer,
+) error {
+	uploadFuture, uploadRelease := bp.p.Upload(ctx, func(p Publications_upload_Params) error {
+		if err := p.SetNs(ns); err != nil {
+			return fmt.Errorf("setting sig: %s", err)
+		}
+
+		if err := p.SetRel(rel); err != nil {
+			return fmt.Errorf("setting sig: %s", err)
+		}
+
+		return nil
+	})
+	defer uploadRelease()
+
+	callback := uploadFuture.Callback()
+
+	buf := make([]byte, 0, 4*1024)
```

Comment:
we're sending 4k bytes. not sure what be an appropriate value

Reply:
yes, we can optimize chunk size later as we have more data. is there a retry mechanism builtin inside Cap'nProto? what if one chunk fails?